src/viadot/orchestration/prefect/utils.py

```diff
@@ -28,6 +28,394 @@
+            "x_units_ago_full_date": r"(\d+)\s+(years?|months?|days?)\s+ago\s+full\s+date",
+            "x_years_ago": r"(\d+)\s+years\s+ago",
+        }
+        self.range_patterns = {
+            "last_x_units": r"last\s+(\d+)\s+(years|months|days)",
+            "y_years_from_x": r"(\d+)\s+years\s+from\s+(\d{4})",
+            "first_x_days_from": r"first\s+(\d+)\s+days\s+from\s+(\w+)\s+(\d{4})",
+            "last_x_days_from": r"last\s+(\d+)\s+days\s+from\s+(\w+)\s+(\d{4})",
+        }
+        self.dynamic_date_format = dynamic_date_format
+        self.dynamic_date_timezone = dynamic_date_timezone
+        self.dynamic_date_symbols = dynamic_date_symbols
+        self.replacements = self.create_date_dict()
+
+    def generate_years(
+        self, last_years: int, from_year: str, num_years: str
+    ) -> List[str]:
+        """
+        Generate a list of years either for the last X years or from a start year.
+
+        Args:
+            last_years (int): The number of years from the current year to include.
+            from_year (str): The starting year.
+            num_years (int): The number of years to generate from the starting year.
+
+        Returns:
+            list: A list of years in ascending order.
+        """
+        current_year = pendulum.now().year
+        if last_years:
+            result = [str(current_year - i) for i in range(last_years)][
+                ::-1
+            ]  # Reversed to ascending order
+            return result
+        elif from_year and num_years:
+            start_year = int(from_year)
+            result = [str(start_year - i) for i in range(num_years)][
+                ::-1
+            ]  # Ascending order
+            return result
+        return []
+
+    def generate_months(self, last_months: int) -> List[str]:
+        """
+        Generate a list of first days of the last X months.
+
+        Args:
+            last_months (int): The number of months to include from the past.
+
+        Returns:
+            list: A list of dates representing the first day of the last X months in ascending order.
+        """
+        current_date = pendulum.now()
+        result = [
+            current_date.subtract(months=i).start_of("month").format("YMM")
+            for i in range(last_months)
+        ][
+            ::-1
+        ]  # Reversed to ascending order
+        return result
+
+    def generate_dates(self, last_days: int) -> List[str]:
+        """
+        Generate a list of dates for the last X days.
+
+        Args:
+            last_days (int): The number of days to include from the past.
+
+        Returns:
+            list: A list of dates in ascending order.
+        """
+        current_date = pendulum.now(self.dynamic_date_timezone)
+        result = [
+            current_date.subtract(days=i).format("YMMDD") for i in range(last_days)
+        ][
+            ::-1
+        ]  # Reversed to ascending order
+        return result
+
+    def process_first_days(
```

Comment:
mark this function internal by adding `_`.
```suggestion
    def _process_first_days(
```